# Project 6 -- Lukas Anell

In [1]:
# /anvil/projects/tdm/data/flights/subset/2006.csv
# /anvil/projects/tdm/data/icecream/combined/products.csv
# /anvil/projects/tdm/data/flights/subset/1997.csv

## Question 1

In [1]:
import pandas as pd

In [2]:
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/2006.csv")

In [3]:
myDF.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
0,2006,1,11,3,743.0,745,1024.0,1018,US,343,...,45,13,0,NaN,0,0,0,0,0,0
1,2006,1,11,3,1053.0,1053,1313.0,1318,US,613,...,27,19,0,NaN,0,0,0,0,0,0
2,2006,1,11,3,1915.0,1915,2110.0,2133,US,617,...,4,11,0,NaN,0,0,0,0,0,0
3,2006,1,11,3,1753.0,1755,1925.0,1933,US,300,...,16,10,0,NaN,0,0,0,0,0,0
4,2006,1,11,3,824.0,832,1015.0,1015,US,765,...,27,12,0,NaN,0,0,0,0,0,0


In [4]:
myDF.shape

(7141922, 29)

In [5]:
pa_flights = myDF[myDF["Dest"] == "PHL"]

In [6]:
pa_flights.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
16,2006,1,11,3,1752.0,1540,1923.0,1654,US,127,...,3,19,0,NaN,0,0,0,149,0,0
17,2006,1,11,3,724.0,730,843.0,845,US,277,...,17,8,0,NaN,0,0,0,0,0,0
18,2006,1,11,3,1153.0,1145,1324.0,1259,US,637,...,3,38,0,NaN,0,0,0,25,0,0
19,2006,1,11,3,528.0,525,640.0,640,US,1658,...,10,10,0,NaN,0,0,0,0,0,0
26,2006,1,11,3,725.0,725,1015.0,1021,US,1636,...,9,9,0,NaN,0,0,0,0,0,0


In [7]:
pa_flights.shape

(108307, 29)

In [8]:
unique_origins = pa_flights[pa_flights['Dest'] == 'PHL']['Origin'].nunique()
unique_origins

66

In [9]:
pa_flights.groupby('Origin')['AirTime'].mean().sort_values(ascending=False)

Origin
SFO    290.109422
OAK    289.584488
LAX    279.706140
SAN    278.428962
SEA    278.121287
          ...    
IAD     33.846006
DCA     33.784762
LGA     33.500000
BWI     31.428748
MDT     26.000000
Name: AirTime, Length: 66, dtype: float64

In [10]:
az_flights = myDF[myDF["Dest"] == "PHX"]
az_flights.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
0,2006,1,11,3,743.0,745,1024.0,1018,US,343,...,45,13,0,NaN,0,0,0,0,0,0
1,2006,1,11,3,1053.0,1053,1313.0,1318,US,613,...,27,19,0,NaN,0,0,0,0,0,0
2,2006,1,11,3,1915.0,1915,2110.0,2133,US,617,...,4,11,0,NaN,0,0,0,0,0,0
3,2006,1,11,3,1753.0,1755,1925.0,1933,US,300,...,16,10,0,NaN,0,0,0,0,0,0
4,2006,1,11,3,824.0,832,1015.0,1015,US,765,...,27,12,0,NaN,0,0,0,0,0,0


In [11]:
az_flights.groupby('Origin')['AirTime'].mean().sort_values(ascending=False)

Origin
HNL    378.044693
LIH    333.385593
KOA    319.184100
OGG    317.876420
PVD    314.596974
          ...    
SAN     50.193324
LAS     45.641012
YUM     40.426352
FLG     40.298725
TUS     26.966643
Name: AirTime, Length: 91, dtype: float64

1.1: 108307 flights

1.2: 290.109422 minutes

1.3: 378.044693 minutes

## Question 2

In [18]:
top_20 = myDF["Dest"].value_counts().head(20).index.tolist()
top_20

['ATL',
 'ORD',
 'DFW',
 'LAX',
 'DEN',
 'IAH',
 'PHX',
 'LAS',
 'EWR',
 'SLC',
 'SFO',
 'BOS',
 'DTW',
 'LGA',
 'CVG',
 'MSP',
 'MCO',
 'CLT',
 'JFK',
 'PHL']

In [19]:
top_dests_df = myDF[myDF["Dest"].isin(top_20)]
top_dests_df.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
0,2006,1,11,3,743.0,745,1024.0,1018,US,343,...,45,13,0,NaN,0,0,0,0,0,0
1,2006,1,11,3,1053.0,1053,1313.0,1318,US,613,...,27,19,0,NaN,0,0,0,0,0,0
2,2006,1,11,3,1915.0,1915,2110.0,2133,US,617,...,4,11,0,NaN,0,0,0,0,0,0
3,2006,1,11,3,1753.0,1755,1925.0,1933,US,300,...,16,10,0,NaN,0,0,0,0,0,0
4,2006,1,11,3,824.0,832,1015.0,1015,US,765,...,27,12,0,NaN,0,0,0,0,0,0


In [23]:
result = top_dests_df.groupby(["Origin", "Dest"])["AirTime"].mean()
result.head()

Origin  Dest
ABE     ATL     99.765806
        CLT     76.243094
        CVG     86.536354
        DTW     75.500000
        JFK     38.750000
Name: AirTime, dtype: float64

I like how easy it is to just chain functions in Python, like in the line for initializing `top_20`. It makes it a lot easier to quickly deal with the data, but I can also see it getting confusing and unreadable if there's too many chained functions.

2.1: See above.

2.2: See above.

2.3: See above.

## Question 3

In [25]:
ice_cream = pd.read_csv("/anvil/projects/tdm/data/icecream/combined/products.csv")
ice_cream.head()

,brand,key,name,subhead,description,rating,rating_count,ingredients
0,bj,0_bj,Salted Caramel Core,Sweet Cream Ice Cream with Blonde Brownies & a...,Find your way to the ultimate ice cream experi...,3.7,208,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),..."
1,bj,1_bj,Netflix & Chilll'd™,Peanut Butter Ice Cream with Sweet & Salty Pre...,There’s something for everyone to watch on Net...,4.0,127,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),..."
2,bj,2_bj,Chip Happens,A Cold Mess of Chocolate Ice Cream with Fudge ...,Sometimes “chip” happens and everything’s a me...,4.7,130,"CREAM, LIQUID SUGAR (SUGAR, WATER), SKIM MILK,..."
3,bj,3_bj,Cannoli,Mascarpone Ice Cream with Fudge-Covered Pastry...,As a Limited Batch that captured the rapture o...,3.6,70,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),..."
4,bj,4_bj,Gimme S’more!™,Toasted Marshmallow Ice Cream with Chocolate C...,It’s a gimme: there’s always room for s’more. ...,4.5,281,"CREAM, SKIM MILK, WATER, LIQUID SUGAR (SUGAR, ..."


In [29]:
ice_cream["rating"].value_counts().head()

rating
4.7    26
4.5    23
4.4    22
4.6    21
4.3    19
Name: count, dtype: int64

In [31]:
ice_cream["rating"].describe()

count    241.000000
mean       4.223237
std        0.620383
min        1.200000
25%        4.000000
50%        4.400000
75%        4.700000
max        5.000000
Name: rating, dtype: float64

In [33]:
print(pd.qcut(ice_cream["rating"], q=4))

0      (1.199, 4.0]
1      (1.199, 4.0]
2        (4.4, 4.7]
3      (1.199, 4.0]
4        (4.4, 4.7]
           ...     
236    (1.199, 4.0]
237      (4.4, 4.7]
238    (1.199, 4.0]
239    (1.199, 4.0]
240    (1.199, 4.0]
Name: rating, Length: 241, dtype: category
Categories (4, interval[float64, right]): [(1.199, 4.0] < (4.0, 4.4] < (4.4, 4.7] < (4.7, 5.0]]


In [30]:
labels = ["Wouldn't Recommend", "Needs Improvement", "Solid Choice", "Fan Favorite"]
ice_cream['rating_phrases'] = pd.qcut(
    ice_cream['rating'], 
    q=4, 
    labels=labels
)
ice_cream.head()

,brand,key,name,subhead,description,rating,rating_count,ingredients,rating_phrases
0,bj,0_bj,Salted Caramel Core,Sweet Cream Ice Cream with Blonde Brownies & a...,Find your way to the ultimate ice cream experi...,3.7,208,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),...",Wouldn't Recommend
1,bj,1_bj,Netflix & Chilll'd™,Peanut Butter Ice Cream with Sweet & Salty Pre...,There’s something for everyone to watch on Net...,4.0,127,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),...",Wouldn't Recommend
2,bj,2_bj,Chip Happens,A Cold Mess of Chocolate Ice Cream with Fudge ...,Sometimes “chip” happens and everything’s a me...,4.7,130,"CREAM, LIQUID SUGAR (SUGAR, WATER), SKIM MILK,...",Solid Choice
3,bj,3_bj,Cannoli,Mascarpone Ice Cream with Fudge-Covered Pastry...,As a Limited Batch that captured the rapture o...,3.6,70,"CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER),...",Wouldn't Recommend
4,bj,4_bj,Gimme S’more!™,Toasted Marshmallow Ice Cream with Chocolate C...,It’s a gimme: there’s always room for s’more. ...,4.5,281,"CREAM, SKIM MILK, WATER, LIQUID SUGAR (SUGAR, ...",Solid Choice


I wasn't sure how to see exactly what the qcut ranges were, so I found the [documentation](https://pandas.pydata.org/docs/reference/api/pandas.qcut.html) for it and saw that I could just print it out.

3.1: See above.

3.2: `[(1.199, 4.0] < (4.0, 4.4] < (4.4, 4.7] < (4.7, 5.0]]` I personally don't think this would be the best way to split up this data, but qcut splits equally so I guess it makes sense.

3.3: See above.

## Question 4

In [34]:
import pandas as pd
my_flights = pd.read_csv('/anvil/projects/tdm/data/flights/subset/1997.csv')

In [35]:
my_flights.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,TaxiIn,TaxiOut,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay
0,1997,1,28,2,1615.0,1615,1728.0,1725,WN,1159,...,4,9,0,NaN,0,NaN,NaN,NaN,NaN,NaN
1,1997,1,29,3,1624.0,1615,1735.0,1725,WN,1159,...,4,11,0,NaN,0,NaN,NaN,NaN,NaN,NaN
2,1997,1,30,4,1626.0,1615,1730.0,1725,WN,1159,...,2,10,0,NaN,0,NaN,NaN,NaN,NaN,NaN
3,1997,1,31,5,1628.0,1615,1729.0,1725,WN,1159,...,4,7,0,NaN,0,NaN,NaN,NaN,NaN,NaN
4,1997,1,14,2,1503.0,1425,1619.0,1535,WN,1221,...,4,8,0,NaN,0,NaN,NaN,NaN,NaN,NaN


In [47]:
bins = [0, 300, 600, 900, 1200, 1500, 1800, 2100, 2400]
labels = ["Late Night", "Early Morning", "Morning", "Late Morning", "Afternoon", "Early Evening", "Evening", "Night"]

my_flights["depart_times"] = pd.cut(my_flights["DepTime"], bins=bins, labels=labels)
my_flights.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay,depart_times,arrival_times
0,1997,1,28,2,1615.0,1615,1728.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
1,1997,1,29,3,1624.0,1615,1735.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
2,1997,1,30,4,1626.0,1615,1730.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
3,1997,1,31,5,1628.0,1615,1729.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
4,1997,1,14,2,1503.0,1425,1619.0,1535,WN,1221,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening


In [46]:
my_flights["depart_times"].value_counts().head()

depart_times
Morning          1046463
Afternoon         998262
Early Evening     997450
Late Morning      968561
Evening           893498
Name: count, dtype: int64

In [48]:
my_flights["arrival_times"] = pd.cut(my_flights["ArrTime"], bins=bins, labels=labels)
my_flights.head()

,Year,Month,DayofMonth,DayOfWeek,DepTime,CRSDepTime,ArrTime,CRSArrTime,UniqueCarrier,FlightNum,...,Cancelled,CancellationCode,Diverted,CarrierDelay,WeatherDelay,NASDelay,SecurityDelay,LateAircraftDelay,depart_times,arrival_times
0,1997,1,28,2,1615.0,1615,1728.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
1,1997,1,29,3,1624.0,1615,1735.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
2,1997,1,30,4,1626.0,1615,1730.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
3,1997,1,31,5,1628.0,1615,1729.0,1725,WN,1159,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening
4,1997,1,14,2,1503.0,1425,1619.0,1535,WN,1221,...,0,NaN,0,NaN,NaN,NaN,NaN,NaN,Early Evening,Early Evening


In [49]:
my_flights["arrival_times"].value_counts().head()

arrival_times
Early Evening    1015952
Evening          1011908
Late Morning      955399
Afternoon         951465
Night             760180
Name: count, dtype: int64

4.1: See above.

4.2: See above.

4.3: Morning is the busiest departure time, early evening is the busiest arrival time.

## Question 5

In [52]:
march = my_flights[my_flights["Month"] == 3]
march["DayOfWeek"].value_counts().sort_index()

DayOfWeek
1    76352
2    61070
3    61093
4    61100
5    61052
6    68031
7    72288
Name: count, dtype: int64

In [53]:
my_flights["DayOfWeek"].value_counts().sort_index()

DayOfWeek
1    790298
2    791617
3    802130
4    785731
5    786342
6    706198
7    749527
Name: count, dtype: int64

In [50]:
my_flights.groupby(["DayOfWeek", "Month"])["AirTime"].sum().unstack()

Month,1,2,3,4,5,6,7,8,9,10,11,12
DayOfWeek,,,,,,,,,,,,
1,5820142.0,5996050.0,7450628.0,6032763.0,5943737.0,7525264.0,6100879.0,6104229.0,7425773.0,5954814.0,5990224.0,7521815.0
2,5894308.0,6014615.0,6036897.0,7438885.0,5973378.0,6023384.0,7587559.0,6088182.0,7471445.0,5977490.0,6012467.0,7485652.0
3,7265554.0,6062085.0,6047810.0,7510912.0,5976554.0,5988220.0,7577076.0,6109375.0,5945379.0,7473880.0,6059871.0,7239442.0
4,7310610.0,6065632.0,5998931.0,6021716.0,7487291.0,6029059.0,7567389.0,6120396.0,5943426.0,7506399.0,5682853.0,5824651.0
5,7415133.0,5992478.0,6028038.0,6061881.0,7528029.0,6046017.0,5840017.0,7627953.0,5915696.0,7477447.0,5761139.0,5989230.0
6,5463380.0,5457435.0,6930980.0,5540389.0,6793947.0,5688313.0,5628626.0,7087879.0,5374278.0,5319835.0,6785410.0,5518159.0
7,5643971.0,5719389.0,7284005.0,5808701.0,5645453.0,7321258.0,5920780.0,7247138.0,5694575.0,5697566.0,7231943.0,5819670.0


In [56]:
mean_airtime = my_flights.groupby(["DayOfWeek", "Month"])["AirTime"].mean().unstack()
mean_airtime

Month,1,2,3,4,5,6,7,8,9,10,11,12
DayOfWeek,,,,,,,,,,,,
1,100.970508,100.466640,100.662397,100.541023,100.238414,100.344881,101.004586,100.411715,99.970019,99.899577,100.911777,102.134739
2,101.115194,100.929907,101.207011,100.080520,99.805815,100.354609,100.721592,100.914669,100.216557,100.175800,101.082144,102.274183
3,102.297167,102.192937,100.902781,100.562492,99.758872,100.454950,100.926753,100.968054,99.828380,100.320537,102.033490,102.805237
4,102.142009,102.097829,100.558720,100.445638,100.072054,100.280413,100.256876,100.632960,99.800615,100.566699,102.355019,102.499754
5,101.678843,101.717414,100.624946,101.235508,99.996400,100.360490,100.459584,100.096488,99.062177,100.079596,103.148246,102.175649
6,103.763960,103.241236,103.319470,103.591590,102.258417,103.110791,102.922506,102.821235,101.735471,102.828549,104.742212,104.909960
7,101.902484,101.355491,101.988309,101.823076,101.125873,101.171257,101.253185,102.014893,100.157855,101.322485,102.244288,103.504962


I thought the dataframes that you get by using .groupby look super cool and also informative. It's cool being able to see this matrix of data about certain days of the week, in certain months.

5.1: As expected, the DayOfWeek value counts for only March are much smaller than the value counts for the whole year.

5.2: See above.

5.3: See above.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
